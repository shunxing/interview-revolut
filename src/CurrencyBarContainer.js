import React from "react";
import { CurrencyContextConsumer } from "./CurrencyContext";
import { CURRENCY_SYMBOL } from "./constants";
import { connect } from "react-redux";

const CurrencyBarComponent = ({ selectedTargetCurrency }) => (
  <CurrencyContextConsumer>
    {context => {
      return context.currencies.map(
        currency =>
          currency.rates[selectedTargetCurrency] &&
          selectedTargetCurrency !== currency.base && (
            <div key={currency.base}>
              {context.base} 1{CURRENCY_SYMBOL[currency.base]} :
              {currency.rates[selectedTargetCurrency]}
              {CURRENCY_SYMBOL[selectedTargetCurrency]}
            </div>
          )
      );
    }}
  </CurrencyContextConsumer>
);

const mapStateToProps = state => ({
  selectedTargetCurrency: state.app.selectedTargetCurrency
});

export const CurrencyBarContainer = connect(mapStateToProps)(
  CurrencyBarComponent
);
