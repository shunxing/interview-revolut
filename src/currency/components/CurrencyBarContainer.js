import React from "react";
import { CurrencyContextConsumer } from "CurrencyContext";
import { CURRENCY_SYMBOL } from "../../constants";
import { connect } from "react-redux";
import "./CurrencyBar.scss";

const CurrencyBarComponent = ({
  selectedTargetCurrency,
  selectedSourceCurrency
}) => (
  <CurrencyContextConsumer>
    {context => {
      const selectedSourceCurrencyContext = context.currencies.find(
        currency => currency.base === selectedSourceCurrency
      );
      return (
        selectedSourceCurrencyContext &&
        selectedSourceCurrencyContext.rates[selectedTargetCurrency] && (
          <div>
            1{CURRENCY_SYMBOL[selectedSourceCurrencyContext.base]} :
            {selectedSourceCurrencyContext.rates[selectedTargetCurrency]}
            {CURRENCY_SYMBOL[selectedTargetCurrency]}
          </div>
        )
      );
    }}
  </CurrencyContextConsumer>
);

const mapStateToProps = state => ({
  selectedTargetCurrency: state.currency.target.value,
  selectedSourceCurrency: state.currency.source.value
});

/* not using default export because it's harder to trace bugs */
export const CurrencyBarContainer = connect(mapStateToProps)(
  CurrencyBarComponent
);
