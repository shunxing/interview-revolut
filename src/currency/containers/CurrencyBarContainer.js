import React from "react";
import { CurrencyContextConsumer } from "CurrencyContext";
import { CURRENCY_SYMBOL } from "../../constants";
import { connect } from "react-redux";
import "./CurrencyBar.scss";
import ShowChartIcon from "@material-ui/icons/ShowChart";
import CompareArrows from "@material-ui/icons/CompareArrows";
import Fab from "@material-ui/core/Fab";

const CurrencyBarComponent = ({
  selectedTargetCurrency,
  selectedSourceCurrency
}) => (
  <CurrencyContextConsumer>
    {context => {
      const selectedSourceCurrencyContext = context.currencies.find(
        currency => currency.base === selectedSourceCurrency
      );
      return (
        <>
          <div className="currency--middle">
            {selectedSourceCurrencyContext &&
              selectedSourceCurrencyContext.rates[selectedTargetCurrency] && (
                <div className="currency__bar">
                  <ShowChartIcon />
                  <div className="currency__bar__text">
                    {`1${
                      CURRENCY_SYMBOL[selectedSourceCurrencyContext.base]
                    } = ${
                      selectedSourceCurrencyContext.rates[
                        selectedTargetCurrency
                      ]
                    } ${CURRENCY_SYMBOL[selectedTargetCurrency]}`}
                  </div>
                </div>
              )}
            <div className="currency__switch-currency">
              <Fab color="secondary">
                <div style={{ transform: "rotate(90deg)" }}>
                  <CompareArrows />
                </div>
              </Fab>
            </div>{" "}
          </div>
        </>
      );
    }}
  </CurrencyContextConsumer>
);

const mapStateToProps = state => ({
  selectedTargetCurrency: state.currency.target.currency,
  selectedSourceCurrency: state.currency.source.currency
});

/* not using default export because it's harder to trace bugs */
export const CurrencyBarContainer = connect(mapStateToProps)(
  CurrencyBarComponent
);
