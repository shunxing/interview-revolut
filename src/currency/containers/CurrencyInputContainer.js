import React, { Component } from "react";
import { connect } from "react-redux";
import { updateCurrencyAmount } from "../redux/currencyActionCreators";
import TextField from "@material-ui/core/TextField";
import { CurrencyContextConsumer } from "../../CurrencyContext";
import { CURRENCY_TYPE } from "../../constants";
import { InputAdornment } from "@material-ui/core";

export class CurrencyInputComponent extends Component {
  render() {
    return (
      <CurrencyContextConsumer>
        {({ currencies }) => (
          <TextField
            value={this.props.currencyAmount}
            options={this.props.availableCurrencies}
            onChange={this.props.onUpdateCurrencyAmount(currencies)}
            placeholder="0"
            InputProps={{
              startAdornment: (
                <InputAdornment position="start">
                  {(this.props.currencyFieldType === CURRENCY_TYPE.SOURCE &&
                    "-") ||
                    (this.props.currencyFieldType === CURRENCY_TYPE.TARGET &&
                      "+")}
                </InputAdornment>
              )
            }}
          />
        )}
      </CurrencyContextConsumer>
    );
  }
}

const mapStateToProps = (state, ownProps) => ({
  currencyAmount: state.currency[ownProps.currencyFieldType].currencyAmount
});

const mapDispatchToProps = (dispatch, ownProps) => ({
  onUpdateCurrencyAmount: currencyRates => event => {
    const currencyAmount = event.target.value;
    const match =
      currencyAmount.match(/^\d+(,\d{3})*(\.\d{1,2})?$/) ||
      currencyAmount.match(/^\d+(,)*(\.)?$/);
    if (match) {
      dispatch(
        updateCurrencyAmount({
          currencyAmount: match[0],
          currencyFieldType: ownProps.currencyFieldType,
          currencyRates: currencyRates
        })
      );
    } else if (currencyAmount === "") {
      dispatch(
        updateCurrencyAmount({
          currencyAmount: currencyAmount,
          currencyFieldType: ownProps.currencyFieldType,
          currencyRates: currencyRates
        })
      );
    }
  }
});

export const CurrencyInputContainer = connect(
  mapStateToProps,
  mapDispatchToProps
)(CurrencyInputComponent);
