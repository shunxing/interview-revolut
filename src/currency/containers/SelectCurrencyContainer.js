import React, { Component } from "react";
import { getSelectedCurrency } from "../../selectors";
import { connect } from "react-redux";
import { selectCurrency } from "../redux/currencyActionCreators";
import { CurrencyContextConsumer } from "CurrencyContext";
import Select from "@material-ui/core/Select";
import MenuItem from "@material-ui/core/MenuItem";
import { CurrenciesMenuDataContextConsumer } from "../../CurrencyContext";

export class SelectCurrencyComponent extends Component {
  render() {
    const { selectedCurrency = "", selectCurrency } = this.props;
    return (
      <CurrencyContextConsumer>
        {({ currencies }) => (
          <CurrenciesMenuDataContextConsumer>
            {({ currenciesMenu }) => {
              return (
                <Select
                  value={selectedCurrency}
                  renderValue={() => selectedCurrency}
                  onChange={selectCurrency(currencies)}
                  MenuProps={{
                    getContentAnchorEl: null,
                    anchorOrigin: {
                      vertical: "bottom",
                      horizontal: "left"
                    }
                  }}
                >
                  {currenciesMenu &&
                    currenciesMenu.selected &&
                    currenciesMenu.selected.map(currency => (
                      <MenuItem value={currency} key={currency}>
                        {currency}
                      </MenuItem>
                    ))}
                </Select>
              );
            }}
          </CurrenciesMenuDataContextConsumer>
        )}
      </CurrencyContextConsumer>
    );
  }
}

const mapStateToProps = (state, ownProps) => ({
  selectedCurrency: getSelectedCurrency(ownProps.currencyFieldType)(state)
});

const mapDispatchToProps = (dispatch, ownProps) => ({
  selectCurrency: currencyRates => event => {
    dispatch(
      selectCurrency({
        selectedCurrency: event.target.value,
        currencyFieldType: ownProps.currencyFieldType,
        currencyRates
      })
    );
  }
});
export const SelectCurrencyContainer = connect(
  mapStateToProps,
  mapDispatchToProps
)(SelectCurrencyComponent);
