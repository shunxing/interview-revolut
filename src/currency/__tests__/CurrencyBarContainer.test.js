/**
 *  HACK TO SET CONTEXT
 * https://gist.github.com/danmakenoise/c503854508f80137fe547943218ccc05
 * https://twitter.com/wesbos/status/976164542140043264
 */
import React from "react";
import { CurrencyBarComponent } from "../containers/CurrencyBarContainer";
import { shallow } from "enzyme";

describe("CurrencyBarComponent", () => {
  let outer = shallow(
      <CurrencyBarComponent
        selectedSourceCurrency="EUR"
        selectedTargetCurrency="USD"
      />
    ),
    ComponentWithContext = outer.props().children;

  it("should not crash when render component", () => {
    shallow(<CurrencyBarComponent />);
  });

  it("should not render currency bar when no currencies found", () => {
    const wrapper = shallow(<ComponentWithContext currencies={[]} />);

    expect(wrapper.find(".currency__bar").exists()).toEqual(false);
    // const wrapper = shallow(ComponentWithContext);
  });

  it("should not render currency bar when no currencies rates in currency found", () => {
    const wrapper = shallow(
      <ComponentWithContext currencies={[{ base: "EUR" }]} />
    );

    expect(wrapper.find(".currency__bar").exists()).toEqual(false);
  });

  it("should render currency bar currency rates found", () => {
    const wrapper = shallow(
      <ComponentWithContext
        currencies={[{ base: "EUR", rates: { USD: "0" } }]}
      />
    );

    expect(wrapper.find(".currency__bar").exists()).toEqual(true);
  });

  it("should display the correct currency rate", () => {
    const wrapper = shallow(
      <ComponentWithContext
        currencies={[{ base: "EUR", rates: { USD: "123" } }]}
      />
    );

    expect(wrapper.find(".currency__bar__text").text()).toEqual("1€ = 123 $");
  });

  it("should display the source currency name when missing symbols", () => {
    outer = shallow(
      <CurrencyBarComponent
        selectedSourceCurrency="CNY"
        selectedTargetCurrency="USD"
      />
    );
    ComponentWithContext = outer.props().children;

    const wrapper = shallow(
      <ComponentWithContext
        currencies={[{ base: "CNY", rates: { USD: "123" } }]}
      />
    );

    expect(wrapper.find(".currency__bar__text").text()).toEqual("1CNY = 123 $");
  });

  it("should display the target currency name when missing symbols", () => {
    outer = shallow(
      <CurrencyBarComponent
        selectedSourceCurrency="EUR"
        selectedTargetCurrency="CNY"
      />
    );
    ComponentWithContext = outer.props().children;
    const wrapper = shallow(
      <ComponentWithContext
        currencies={[{ base: "EUR", rates: { CNY: "123" } }]}
      />
    );

    expect(wrapper.find(".currency__bar__text").text()).toEqual("1€ = 123 CNY");
  });

  it("should display the source and target currency name when missing symbols", () => {
    outer = shallow(
      <CurrencyBarComponent
        selectedSourceCurrency="SGD"
        selectedTargetCurrency="CNY"
      />
    );

    ComponentWithContext = outer.props().children;
    const wrapper = shallow(
      <ComponentWithContext
        currencies={[{ base: "SGD", rates: { CNY: "123" } }]}
      />
    );

    expect(wrapper.find(".currency__bar__text").text()).toEqual(
      "1SGD = 123 CNY"
    );
  });

  it("should match snapshot", () => {
    const wrapper = shallow(
      <ComponentWithContext
        currencies={[{ base: "EUR", rates: { USD: "123" } }]}
      />
    );

    expect(wrapper).toMatchSnapshot();
  });
});

/**
 * no testing for connect part because it's not part of unit tests
 * https://hackernoon.com/unit-testing-redux-connected-components-692fa3c4441c
 */
