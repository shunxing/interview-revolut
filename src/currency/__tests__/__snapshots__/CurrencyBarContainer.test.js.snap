// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CurrencyBarComponent should match snapshot 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Unknown
    currencies={
      Array [
        Object {
          "base": "EUR",
          "rates": Object {
            "USD": "123",
          },
        },
      ]
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "children": <div
        className="currency--middle"
      >
        <div
          className="currency__switch-currency"
        >
          <WithStyles(Fab)
            color="secondary"
          >
            <div
              style={
                Object {
                  "transform": "rotate(90deg)",
                }
              }
            >
              <pure(CompareArrowsIcon) />
            </div>
          </WithStyles(Fab)>
        </div>
      </div>,
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          undefined,
          <div
            className="currency__switch-currency"
          >
            <WithStyles(Fab)
              color="secondary"
            >
              <div
                style={
                  Object {
                    "transform": "rotate(90deg)",
                  }
                }
              >
                <pure(CompareArrowsIcon) />
              </div>
            </WithStyles(Fab)>
          </div>,
        ],
        "className": "currency--middle",
      },
      "ref": null,
      "rendered": Array [
        undefined,
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": <WithStyles(Fab)
              color="secondary"
            >
              <div
                style={
                  Object {
                    "transform": "rotate(90deg)",
                  }
                }
              >
                <pure(CompareArrowsIcon) />
              </div>
            </WithStyles(Fab)>,
            "className": "currency__switch-currency",
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "children": <div
                style={
                  Object {
                    "transform": "rotate(90deg)",
                  }
                }
              >
                <pure(CompareArrowsIcon) />
              </div>,
              "color": "secondary",
              "onClick": undefined,
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": <pure(CompareArrowsIcon) />,
                "style": Object {
                  "transform": "rotate(90deg)",
                },
              },
              "ref": null,
              "rendered": Object {
                "instance": null,
                "key": undefined,
                "nodeType": "class",
                "props": Object {},
                "ref": null,
                "rendered": null,
                "type": [Function],
              },
              "type": "div",
            },
            "type": [Function],
          },
          "type": "div",
        },
      ],
      "type": "div",
    },
    "type": Symbol(react.fragment),
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "children": <div
          className="currency--middle"
        >
          <div
            className="currency__switch-currency"
          >
            <WithStyles(Fab)
              color="secondary"
            >
              <div
                style={
                  Object {
                    "transform": "rotate(90deg)",
                  }
                }
              >
                <pure(CompareArrowsIcon) />
              </div>
            </WithStyles(Fab)>
          </div>
        </div>,
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            undefined,
            <div
              className="currency__switch-currency"
            >
              <WithStyles(Fab)
                color="secondary"
              >
                <div
                  style={
                    Object {
                      "transform": "rotate(90deg)",
                    }
                  }
                >
                  <pure(CompareArrowsIcon) />
                </div>
              </WithStyles(Fab)>
            </div>,
          ],
          "className": "currency--middle",
        },
        "ref": null,
        "rendered": Array [
          undefined,
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": <WithStyles(Fab)
                color="secondary"
              >
                <div
                  style={
                    Object {
                      "transform": "rotate(90deg)",
                    }
                  }
                >
                  <pure(CompareArrowsIcon) />
                </div>
              </WithStyles(Fab)>,
              "className": "currency__switch-currency",
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "children": <div
                  style={
                    Object {
                      "transform": "rotate(90deg)",
                    }
                  }
                >
                  <pure(CompareArrowsIcon) />
                </div>,
                "color": "secondary",
                "onClick": undefined,
              },
              "ref": null,
              "rendered": Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": <pure(CompareArrowsIcon) />,
                  "style": Object {
                    "transform": "rotate(90deg)",
                  },
                },
                "ref": null,
                "rendered": Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "class",
                  "props": Object {},
                  "ref": null,
                  "rendered": null,
                  "type": [Function],
                },
                "type": "div",
              },
              "type": [Function],
            },
            "type": "div",
          },
        ],
        "type": "div",
      },
      "type": Symbol(react.fragment),
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
